### PR TITLE
Add unit tests for group validation webhook OSD-3361

### PR DESCRIPTION
**Pull request by csheremeta**
Name                                          Stmts   Miss  Cover   Missing
---------------------------------------------------------------------------
src/webhook/group_validation.py                  55     17    69%   22-40, 55, 59, 68

### CODE REVIEW

**Comment by csheremeta** [comment]
now:
src/webhook/group_validation.py                  64     17    73%   25-43, 69, 75, 87
(thanks @jewzaam )

**Comment by jewzaam** [comment]
Since I applied a pretty big change on top of this in https://github.com/csheremeta/managed-cluster-validating-webhooks/pull/1 I abstain from review.

Note the layered-cs-sre-admins stuff partially addresses https://issues.redhat.com/browse/OSD-2944 and I'll wait to do any more tweaks until this PR lands.

File: src/webhook/group_validation.py [hunk 1]
```diff
@@ -37,34 +40,54 @@ def handle_request():
...
+    userinfo = {}
+    if 'userInfo' in body_dict:
+      userinfo = body_dict['userInfo']
+    user_groups = []
+    if 'groups' in userinfo:
+      user_groups = userinfo['groups']
+
+    user_name = None
+    if 'username' in userinfo:
+      user_name = userinfo['username']
```
[the remark this answers is not in the document]
Comment (jewzaam) [reply src/webhook/group_validation.py:64]: Yea, should be replaced.  I did this then updated validation a little but didn't want to go too far.  We could probably validate that a username exists and just trust if we get this far it's ok.

File: src/webhook/namespace_validation.py [hunk 2]
```diff
@@ -40,7 +40,12 @@ def handle_request():
 def get_response(req, debug=False):
   try:
     body_dict = req.json['request']
-    requester_group_memberships = body_dict['userInfo']['groups']
+    requester_group_memberships = []
+    if 'userInfo' in body_dict:
+      if 'groups' in body_dict['userInfo']:
+        requester_group_memberships = body_dict['userInfo']['groups']
+    else:
+      return responses.response_invalid()
```
[the remark this answers is not in the document]
Comment (jewzaam) [reply src/webhook/namespace_validation.py:48]: Yes, it's necessary and was causing tests to fail.  There was some weirdness with tests using the same names for global scoped vars that was masking this until I moved the group validation test global vars into a class (and then namespace validation test global vars).

File: README.md [hunk 3]
```diff
@@ -8,7 +8,7 @@ A Flask app designed to act as a webhook admission controller for OpenShift.
 
 Configuration for this webhook is provided by environment variables:
 
-* `GROUP_VALIDATION_PREFIX` - Group prefix to apply the webhook, such as `osd-` to apply to `CREATE`, `UPDATE`, `DELETE` operations on groups starting with `osd-`. (default: `osd-sre-`)
+* `GROUP_VALIDATION_PROTECTED_GROUP_REGEX` - Regular expression for protected group names, such as `osd-` to apply to `CREATE`, `UPDATE`, `DELETE` operations on groups starting with `osd-`. (default: `(^osd-sre.*|^dedicated-admins$|^cluster-admins$|^layered-cs-sre-admins$)`)
```
[the remark this answers is not in the document]
Comment (jewzaam) [reply README.md:11]:
```suggestion
* `GROUP_VALIDATION_PROTECTED_GROUP_REGEX` - Regular expression for groups that cannot be edited by non-SRE users, such as `osd-` to apply to `CREATE`, `UPDATE`, `DELETE` operations on groups starting with `osd-`. (default: `(^osd-sre.*|^dedicated-admins$|^cluster-admins$|^layered-cs-sre-admins$)`)
```

**Comment by jewzaam** [comment]
So, I could merge one a suggestion and did but it's not my PR.. sorry!  Didn't know I could do that.

**Comment by csheremeta** [comment]
/hold cancel
@jewzaam @lisa I added more test cases to fully cover the logic introduced by Naveen's commits, PTAL

**Comment by csheremeta** [comment]
@lisa I don't see value in adding a create_create_request function. the only reason I separated out the create_request from create_delete_request is because the delete requests look at the oldObject metadata, instead of the object metadata used by create/update/etc requests. the string represented in either function's "operation" tag could be "bobsyouruncle" for all it matters - this difference was not the point of the exercise

**Comment by jewzaam** [comment]
@csheremeta at this time the operation doesn't matter to the code so I'm in agreement.  The only benefit I can see to adding a "create" will be future proof for the unit tests.  My recommendation is to not hold up this PR but to create a new jira card for adding "create" vs "update" tests for all the unit tests in this repo.  I know we don't have it for everything, so rather it's covered for everything.

**Comment by jewzaam** [comment]
Looks reasonable to me, will let someone else +1 it though given my PR into this, so if @lisa could review it would be great.

/approve

File: src/webhook/group_validation.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lisa) [reply src/webhook/group_validation.py:64]: `userInfo` [is a part of the required spec](https://github.com/kubernetes/api/blob/master/admission/v1/types.go#L94-L95) for an `AdmissionReview`, but within the `UserInfo` struct, everything (including the `username`) [is optional](https://github.com/kubernetes/api/blob/master/authentication/v1/types.go#L100-L117). So checking for the username and groups is probably where we want to be.

File: src/webhook/test_namespace_validation.py [hunk 4]
```diff
@@ -4,52 +4,6 @@
 from webhook import namespace_validation
 
 
-PRIVILEGED_NAMESPACES = (
```
Comment (lisa) [inline src/webhook/test_namespace_validation.py:7]: Why are the namespace tests changing in a PR about group validation tests? Will these changes introduce merge conflicts with https://github.com/openshift/managed-cluster-validating-webhooks/pull/66 (cc @jewzaam)?

File: src/webhook/test_namespace_validation.py [hunk 5]
```diff
@@ -67,6 +21,51 @@ class FakeRequest(object):
 
 
 class TestNamespaceValidation(unittest.TestCase):
+    PRIVILEGED_NAMESPACES = (
+        "kube-admin",
+        "kube-foo",
+        "kube-",
```
Comment (lisa) [inline src/webhook/test_namespace_validation.py:27]:
```suggestion
```

To be in line with changes in https://github.com/openshift/managed-cluster-validating-webhooks/pull/66, I'll make the same suggested changes here.

`kube-` is not a valid Namespace name, and so it doesn't make sense to test it. The Kubernetes API would deny the creation of such a Namespace.

File: src/webhook/test_namespace_validation.py [hunk 6]
```diff
@@ -67,6 +21,51 @@ class FakeRequest(object):
...
+        "openshift",
+        "openshifter",
+        "openshift-foo",
+        "ops-health-monitoring",
+        "management-infra",
+        "default",
+        "logging",
+        "sre-app-check",
+        "redhat-user",
+        "redhat-",
```
Comment (lisa) [inline src/webhook/test_namespace_validation.py:37]: To be in line with changes in https://github.com/openshift/managed-cluster-validating-webhooks/pull/66, I'll make the same suggested changes here.

`redhat-` is not a valid Namespace name, and so it doesn't make sense to test it. The Kubernetes API would deny the creation of such a Namespace.
```suggestion
```

File: src/webhook/test_namespace_validation.py (hunk 5 above)
Comment (csheremeta) [reply src/webhook/test_namespace_validation.py]: done

File: src/webhook/test_namespace_validation.py (hunk 6 above)
Comment (csheremeta) [reply src/webhook/test_namespace_validation.py]: done

**Pull request merged**